### PR TITLE
feat: add CPU offloading support for gradient checkpointing

### PR DESCRIPTION
**Pull request by kohya-ss**
This pull request introduces support for activation CPU offloading during gradient checkpointing across several model classes, reducing VRAM usage and enabling larger data (latents) training. The main changes include adding an `activation_cpu_offloading` flag to model classes, updating the gradient checkpointing logic to optionally wrap forward functions for CPU offloading, and exposing the new option via the training script CLI. Additionally, device consistency is ensured after offloading operations to prevent device mismatch errors.

This is particularly effective when the latents tensor is long, meaning the image size is large or the number of frames is large. Therefore, it is effective for training with videos, but its effect is limited when training with still images. However, it allows you to increase the batch size.

When enabled, VRAM usage will decrease but main RAM usage will increase. The training speed will be slightly slower. I think it will be between a few percent and 20% depending on the model and environment.

The changes in memory usage in preliminary studies are as follows, with `--fp8_base` and `--fp8_scaled` (except HunyuanVideo) enabled.

|Model|data|no cpu offload|cpu offload|
|---|---|---|---|
|HunyuanVideo|1280x720@41f|42.7GB|31.9GB|
|Wan2.2|1280x720@25f|46.4GB|34.4GB|
|FramePack 1 frame|1024x1024 bs=3|43.5GB|35.5GB|
|FLUX.1 Kontext|1024x1024 bs=4|30.5GB|28.5GB|
|Qwen-Image|640x640 bs=16|42.3GB|33.9GB|

To enable this feature, specify `--gradient_checkpointing_cpu_offload` in addition to `--gradient_checkpointing`.

We have confirmed that the loss decreases similarly to when the option is not enabled, but the training results have not yet been confirmed.

~~We have not tested whether it works with block swap or multi-GPU training.~~
It seems to work together with block swap. Multi-GPU training is not tested yet.

Feedback is welcome.

### CODE REVIEW

**Comment by FurkanGozukara** [comment]
if this doesn't work with block swapping it useless isnt it? but otherwise looking nice

i will test now

**Comment by FurkanGozukara** [comment]
ok i just tested and sadly i didn't reduce my 14500 MB lowest VRAM config

58 block swap, 128 rank qwen lora and 1328px

 
image

**Comment by kohya-ss** [comment]
This option is most effective when training with large latent sizes (i.e., videos with many frames) or when using a large batch size.

Additional benchmark results for 1024x1024 Qwen-Image LoRA training with various settings (rank (dim)=4):

|batch size|blocks to swap|CPU offload|GPU memory|time/it|
|-|-|-|-|-|
|1|0|No|29.5GB|7.23s|
|1|0|Yes|28.1GB|7.73s|
|1|16|No|23.1GB|7.87s|
|1|13|Yes|23.0GB|8.19s|
|8|58|No|23.0GB|54.00s|
|8|33|Yes|23.4GB|56.19s|
|12|45|Yes|23.5GB|84.66s|
|4|50|Yes|10.5GB|30.67s|

Run on 48GB A6000 GPU (Power Limit 60%) with PyTorch 2.6.0 and flash attention 2.7.4.post1 with CUDA 12.4.

**Comment by Sarania** [comment]
Oh nice I missed this one when it dropped, wondered about the code I was seeing when merging it. Nice little add, anything to help us tame these beasts is a win!
